lib/etl.ex

```diff
@@ -4,67 +4,144 @@ defmodule Etl do
             pids: [],
             subscriptions: []
 
-  @type run_opts :: [
-          source: Etl.Source.t(),
-          transformations: [Etl.Transformation.t()],
-          destination: Etl.Destination.t(),
-          dictionary: Etl.Dictionary.t(),
-          error_handler: (event :: term(), reason :: term() -> no_return()),
-          max_demand: pos_integer(),
-          min_demand: pos_integer()
-        ]
-
-  @spec run(run_opts) :: t
-  def run(opts) do
-    source = Keyword.fetch!(opts, :source)
-    destination = Keyword.fetch!(opts, :destination)
-    dictionary = Keyword.get(opts, :dictionary)
-    transformations = Keyword.get(opts, :transformations, [])
-
-    error_handler =
-      Keyword.get(opts, :error_handler, fn event, reason ->
-        Logger.debug(fn ->
-          "Event #{inspect(event)} has been evicted, reason: #{inspect(reason)}"
+  defmodule Pipeline do
+    defstruct graph: nil, context: nil
+
+    def new(opts \\ []) do
+      %__MODULE__{
+        graph: Graph.new(type: :directed),
+        context: %Etl.Context{
+          min_demand: Keyword.get(opts, :min_demand, 500),
+          max_demand: Keyword.get(opts, :max_demand, 1000),
+          dynamic_supervisor: Keyword.get(opts, :dynamic_supervisor, Etl.DynamicSupervisor)
+        }
+      }
+    end
+
+    def heads(pipeline) do
+      pipeline.graph
+      |> Graph.vertices()
+      |> Enum.filter(fn v -> Graph.in_degree(pipeline.graph, v) == 0 end)
+    end
+
+    def tails(pipeline) do
+      pipeline.graph
+      |> Graph.vertices()
+      |> Enum.filter(fn v -> Graph.out_degree(pipeline.graph, v) == 0 end)
+    end
+
+    def update_graph(pipeline, fun) do
+      Map.update!(pipeline, :graph, fun)
+    end
+
+    def label_stage(pipeline, stage, label) do
+      update_graph(pipeline, fn g ->
+        Graph.label_vertex(g, stage, label)
+      end)
+    end
+  end
+
+  def pipeline(stage, opts \\ []) do
+    Pipeline.new(opts)
+    |> to(stage)
+  end
+
+  def to(pipeline, stage) do
+    case Pipeline.tails(pipeline) do
+      [] ->
+        child_spec = to_child_spec(stage, pipeline)
+        Pipeline.update_graph(pipeline, &Graph.add_vertex(&1, child_spec))
+
+      tails ->
+        Enum.reduce(tails, pipeline, fn tail, pipeline ->
+          Pipeline.update_graph(pipeline, fn g ->
+            case Graph.vertex_labels(g, tail) |> Keyword.get(:dispatcher) do
+              {GenStage.PartitionDispatcher, opts} ->
+                partitions = Keyword.fetch!(opts, :partitions)
+
+                partitions
+                |> Enum.zip(to_child_spec(stage, pipeline, Enum.count(partitions)))
+                |> Enum.reduce(g, fn {partition, child_spec}, g ->
+                  g
+                  |> Graph.add_vertex(child_spec, subscription_opts: [partition: partition])
+                  |> Graph.add_edge(tail, child_spec)
+                end)
+
+              _ ->
+                Graph.add_edge(g, tail, to_child_spec(stage, pipeline))
```

Comment:
break this into separate function(s)

Reply:
and/or rename the `g` variable so it's easier to follow through this series of reducers.